engine/src/main/java/org/camunda/bpm/engine/impl/cmd/StartProcessInstanceByMessageAndProcessDefinitionIdCmd.java

```diff
@@ -0,0 +1,98 @@
+import org.camunda.bpm.engine.impl.persistence.deploy.DeploymentCache;
+import org.camunda.bpm.engine.impl.persistence.entity.ExecutionEntity;
+import org.camunda.bpm.engine.impl.persistence.entity.ProcessDefinitionEntity;
+import org.camunda.bpm.engine.impl.pvm.process.ActivityImpl;
+import org.camunda.bpm.engine.runtime.ProcessInstance;
+import org.camunda.bpm.model.bpmn.BpmnModelInstance;
+import org.camunda.bpm.model.bpmn.instance.MessageEventDefinition;
+import org.camunda.bpm.model.bpmn.instance.StartEvent;
+
+import static org.camunda.bpm.engine.impl.util.EnsureUtil.ensureNotNull;
+
+/**
+ * Command to start a process instance by message. 
+ * 
+ * See {@link StartProcessInstanceByMessageCmd}. In addition, this command allows
+ * specifying the exactly version of the process definition with the given id.
+ * 
+ * @see CAM-2517
+ * 
+ * @author Philipp Ossler
+ */
+public class StartProcessInstanceByMessageAndProcessDefinitionIdCmd implements Command<ProcessInstance> {
+
+  protected final String messageName;
+  protected final String businessKey;
+  protected final Map<String, Object> processVariables;
+  protected final String processDefinitionId;
+
+  public StartProcessInstanceByMessageAndProcessDefinitionIdCmd(String messageName, String processDefinitionId, String businessKey, Map<String, Object> processVariables) {
+    this.messageName = messageName;
+    this.businessKey = businessKey;
+    this.processVariables = processVariables;    
+    this.processDefinitionId = processDefinitionId;
+  }
+
+  public ProcessInstance execute(CommandContext commandContext) {
+    ensureNotNull("Cannot start process instance by message and process definition id", "messageName", messageName);
+    ensureNotNull("Cannot start process instance by message and process definition id", "processDefinitionId", processDefinitionId);
+
+    DeploymentCache deploymentCache = Context
+      .getProcessEngineConfiguration()
+      .getDeploymentCache();
+
+    ProcessDefinitionEntity processDefinition = deploymentCache.findDeployedProcessDefinitionById(processDefinitionId);
+    ensureNotNull("No process definition found for id '" + processDefinitionId + "'", "processDefinition", processDefinition);
+
+    String activityId = findStartActivityIdByMessage(messageName);
+    ensureNotNull("Cannot start process instance by message: no message start event with name '" + messageName + "' found for process definition with id '" + processDefinitionId + "'", "activityId", activityId);
+    
+    ActivityImpl startActivity = processDefinition.findActivity(activityId);
+    ExecutionEntity processInstance = processDefinition.createProcessInstance(businessKey, startActivity);
+
+    if (processVariables != null) {
+      processInstance.setVariables(processVariables);
+    }
+
+    processInstance.start();
+
+    return processInstance;
+  }
+
+  private String findStartActivityIdByMessage(String messageName) {
```

Comment:
While this is valid, I would propose to iterate the message declarations added to the ProcessDefinitionScope by the Parser/Deployer. Something along the lines of:

```java
String activityIdForMessageName = null;
for (EventSubscriptionDeclaration declaration : EventSubscriptionDeclaration.getDeclarationsForScope(subProcessInstance.getScopeActivity())) {
  if(MessageEventHandler.EVENT_HANDLER_TYPE.equals(declaration.getEventType()) && messageName.equals(declaration.getEventName())) {
    activityIdForMessageName = declaration.getActivityId();
  }
}
```

The parser will preprocess the information from the model and extract the executable aspects.